conformance: move Gateway API conformance test logic to Make target

pillai-ashwin opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title, description and a `Fixes: #XXX` line if the commit addresses a particular GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer's Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

## Description of change

Move the Gateway API conformance test execution logic from the GitHub workflow into a reusable `gateway-api-conformance` Make target. This change:

- **Consolidates logic**: Moves duplicated `go test` command logic from workflow YAML to Makefile
- **Preserves behavior**: Maintains exact original behavior (`go test | tparse -progress`) with no functional changes
- **Improves maintainability**: Eliminates code duplication and centralizes test execution logic
- **Prepares for future enhancement**: Creates foundation for integrating with `GOTEST_FORMATTER` from `Makefile.defs` in a subsequent PR

The workflow now calls `make gateway-api-conformance` instead of executing the test pipeline inline, making it consistent with other test targets in the repository.

This is **Step 1** of a two-step approach as requested in [PR review feedback](https://github.com/cilium/cilium/pull/41038#discussion_r2268096570). 
Step 2 (separate PR) will add owner attribution by integrating with `GOTEST_FORMATTER`.

Fixes: #41029

```release-note
ci: Gateway API conformance test logic moved to reusable Make target for better maintainability.
```

sayboras commented on .github/workflows/conformance-gateway-api.yaml:
```diff
@@ -311,6 +319,9 @@ jobs:
               -json \
+            | tee \
+              >(go-junit-report -parser gojson -code-owners="${CILIUM_CLI_CODE_OWNERS_PATHS}" -code-owners-prefix github.com/cilium/cilium/ -out "${JUNIT_PATH}") \
+              >(go run ./tools/testowners --code-owners="${CILIUM_CLI_CODE_OWNERS_PATHS}") \
```
tbh, I am not sure how it works. I just triggered the test and hope it fails to see the result 😓

joestringer replied:
This workflow uses `workflow_dispatch`, so PRs from outside the Cilium organization will not be tested directly. When you post `/test`, Ariane observes this trigger and decides how to safely execute the workflow while taking into account user access to secrets and such. In order to protect against abuse, Ariane either:

1. PR is outside Cilium org - Trigger a call to the workflow checked into the `main` branch
2. PR is inside Cilium org - Trigger a call to the workflow checked into the branch from the PR.

So in order to test we'll need to make a copy of the PR inside the org and trigger the tests there.

sayboras replied:
ah thanks, I missed the workflow_dispatch setting.

sayboras reviewed (approved):
Thanks, looks good to me, at least for the changes didn't break if the conformance test is running well.

joestringer commented on .github/workflows/conformance-gateway-api.yaml:
```diff
@@ -327,6 +338,9 @@ jobs:
               -json \
+            | tee \
+              >(go-junit-report -parser gojson -code-owners="${CILIUM_CLI_CODE_OWNERS_PATHS}" -code-owners-prefix github.com/cilium/cilium/ -out "${JUNIT_PATH}") \
+              >(go run ./tools/testowners --code-owners="${CILIUM_CLI_CODE_OWNERS_PATHS}") \
```
It would be ideal if we could reuse the `GOTEST_FORMATTER` variable directly from `Makefile.defs`, as there's a lot of nuance about exactly how these commands get executed, how to ensure the output completes to each subcommand, and so on. We've hit weird and difficult-to-debug issues when these steps are not done in the particular way that the Makefile does them.

Maybe it's time to move the core logic of this step of the workflow into a script or Makefile target so that we can reuse the same logic. I would propose that assuming @cilium/sig-servicemesh and @cilium/github-sec are good with it that we split this effort into two steps: (1) First move the content of `Run Gateway API conformance test` to a script or Make target, then (2) a separate commit or even PR that ensures the `go test` flags are correct for code owner attribution + pipe the output into `GOTEST_FORMATTER` in a similar way that the other Make targets do.

pillai-ashwin replied:
Thanks for the detailed feedback @joestringer! 

I've implemented **Step 1** (for this PR):
- Moved the Gateway API conformance test logic from the GitHub workflow into a `gateway-api-conformance` Make target
- Preserved the original behavior exactly (`go test | tparse -progress` with no owner attribution)
- Consolidated the logic in one place to avoid maintenance issues
- Workflow will now call the Make target instead of duplicating the pipeline inline

Next Steps (another PR)

**Step 2** will replace the manual `tparse -progress` pipeline with `$(GOTEST_FORMATTER)` from `Makefile.defs`, which will:
- Add owner attribution automatically when `LOG_CODEOWNERS=1` is set
- Leverage the established and tested pipeline logic mentioned

joestringer replied:
OK, thanks. If you have changed fundamentally what this PR is doing, please ensure that the PR title and description match the current proposal. If there is another PR that requires this PR, please ensure it is marked draft until we merge this PR; otherwise GitHub will likely ask the reviewers to review both PRs and it may end up duplicating the efforts that reviewers put into the PRs.

pillai-ashwin replied:
Done!

joestringer reviewed (changes requested):
I think that duplicating the output steps here vs. the Makefile will likely introduce bugs and make it harder to maintain these steps going forward. I'd rather we consolidate on using the existing definitions of how to invoke go-junit-report and tparse from `Makefile.defs`.

pillai-ashwin committed:
conformance: move Gateway API conformance test logic to Make target

Move the Gateway API conformance test execution from the GitHub workflow
into a Make target to consolidate the logic in one place.

We avoid duplicating test output pipeline
logic and prepare for reusing the established GOTEST_FORMATTER from
Makefile.defs in a future change.

Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

joestringer reviewed (approved):
LGTM. We probably need to clone this PR to cilium/cilium to validate the change. Perhaps @cilium/sig-servicemesh could help with that given this is a gateway API change.
